From a48446263288e89310fed3a5ea4041f4b4add28b Mon Sep 17 00:00:00 2001
From: Matthias Schoepfer <m.schoepfer@rethinkrobotics.com>
Date: Fri, 22 Sep 2023 18:31:27 +0200
Subject: [PATCH] rosidl_generator_py_generate_interfaces: use cmake targets

Another approach to use cmake targets instead of old stlye variables
that will leak host path into target.

Signed-off-by: Matthias Schoepfer <m.schoepfer@rethinkrobotics.com>

%% original patch: 0001-rosidl_generator_py_generate_interfaces-use-cmake-ta.patch

%% original patch: 0001-rosidl_generator_py_generate_interfaces-use-cmake-ta.patch

%% original patch: 0001-rosidl_generator_py_generate_interfaces-use-cmake-ta.patch

%% original patch: 0001-rosidl_generator_py_generate_interfaces-use-cmake-ta.patch

%% original patch: 0001-rosidl_generator_py_generate_interfaces-use-cmake-ta.patch
---
 ...idl_generator_py_generate_interfaces.cmake | 47 ++++---------------
 1 file changed, 9 insertions(+), 38 deletions(-)

diff --git a/cmake/rosidl_generator_py_generate_interfaces.cmake b/cmake/rosidl_generator_py_generate_interfaces.cmake
index a6d8263..597f8a0 100644
--- a/cmake/rosidl_generator_py_generate_interfaces.cmake
+++ b/cmake/rosidl_generator_py_generate_interfaces.cmake
@@ -17,11 +17,8 @@ find_package(rosidl_runtime_c REQUIRED)
 find_package(rosidl_typesupport_c REQUIRED)
 find_package(rosidl_typesupport_interface REQUIRED)
 
-find_package(PythonInterp 3.6 REQUIRED)
-
-find_package(python_cmake_module REQUIRED)
-find_package(PythonExtra MODULE REQUIRED)
-
+find_package(Python3 REQUIRED COMPONENTS Development NumPy)
+message(WARNING "DEBUG DA FUCK: Python3_NumPy_FOUND = ${Python3_NumPy_FOUND} Python3_NumPy_INCLUDE_DIRS = ${Python3_NumPy_INCLUDE_DIRS} Python3_EXECUTABLE = ${Python3_EXECUTABLE} Python3_INCLUDE_DIRS = ${Python3_INCLUDE_DIRS}" )
 # Get a list of typesupport implementations from valid rmw implementations.
 rosidl_generator_py_get_typesupports(_typesupport_impls)
 
@@ -126,7 +123,7 @@ endif()
 
 set(_target_suffix "__py")
 
-set(_PYTHON_EXECUTABLE ${PYTHON_EXECUTABLE})
+set(PYTHON_EXECUTABLE ${Python3_EXECUTABLE})
 if(WIN32 AND "${CMAKE_BUILD_TYPE}" STREQUAL "Debug")
   set(PYTHON_EXECUTABLE ${PYTHON_EXECUTABLE_DEBUG})
 endif()
@@ -148,8 +145,8 @@ macro(set_properties _build_type)
     PREFIX ""
     LIBRARY_OUTPUT_DIRECTORY${_build_type} ${_output_path}
     RUNTIME_OUTPUT_DIRECTORY${_build_type} ${_output_path}
-    OUTPUT_NAME "${PROJECT_NAME}_s__${_typesupport_impl}${PythonExtra_EXTENSION_SUFFIX}"
-    SUFFIX "${PythonExtra_EXTENSION_EXTENSION}")
+    OUTPUT_NAME "${PROJECT_NAME}_s__${_typesupport_impl}.${Python3_SOABI}"
+    SUFFIX "${Python3_EXTENSION}")
 endmacro()
 
 macro(set_lib_properties _build_type)
@@ -174,40 +171,15 @@ add_dependencies(
 
 target_link_libraries(
   ${_target_name_lib}
-  ${PythonExtra_LIBRARIES}
+  Python3::Python
+  Python3::NumPy
 )
 target_include_directories(${_target_name_lib}
   PRIVATE
   ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_c
   ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_py
-  ${PythonExtra_INCLUDE_DIRS}
 )
 
-# Check if numpy is in the include path
-find_file(_numpy_h numpy/numpyconfig.h
-  PATHS ${PythonExtra_INCLUDE_DIRS}
-)
-
-if(APPLE OR WIN32 OR NOT _numpy_h)
-  # add include directory for numpy headers
-  set(_python_code
-    "import numpy"
-    "print(numpy.get_include())"
-  )
-  execute_process(
-    COMMAND "${PYTHON_EXECUTABLE}" "-c" "${_python_code}"
-    OUTPUT_VARIABLE _output
-    RESULT_VARIABLE _result
-    OUTPUT_STRIP_TRAILING_WHITESPACE
-  )
-  if(NOT _result EQUAL 0)
-    message(FATAL_ERROR
-      "execute_process(${PYTHON_EXECUTABLE} -c '${_python_code}') returned "
-      "error code ${_result}")
-  endif()
-  message(STATUS "Using numpy include directory: ${_output}")
-  target_include_directories(${_target_name_lib} PUBLIC "${_output}")
-endif()
 
 rosidl_get_typesupport_target(c_typesupport_target "${rosidl_generate_interfaces_TARGET}" "rosidl_typesupport_c")
 target_link_libraries(${_target_name_lib} ${c_typesupport_target})
@@ -246,7 +218,8 @@ foreach(_typesupport_impl ${_typesupport_impls})
   target_link_libraries(
     ${_target_name}
     ${_target_name_lib}
-    ${PythonExtra_LIBRARIES}
+    Python3::Python
+    Python3::NumPy
     ${rosidl_generate_interfaces_TARGET}__${_typesupport_impl}
   )
 
@@ -254,9 +227,7 @@ foreach(_typesupport_impl ${_typesupport_impls})
     PUBLIC
     ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_c
     ${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_py
-    ${PythonExtra_INCLUDE_DIRS}
   )
-
   target_link_libraries(${_target_name} ${c_typesupport_target})
 
   ament_target_dependencies(${_target_name}
-- 
2.41.0

